cilium/CFP-30788-managing-hubble-metrics-cardinality.md

```diff
@@ -0,0 +1,135 @@
+
+- When the cluster is using a managed prometheus offering where the prometheus configuration is managed by the cloud provider and custom ingestion rules cannot be applied.
+- When a high scale cluster is generating high cardinality of metrics putting pressure on the prometheus ingestion, storage and query performance.
+- When the cluster admin wants to dynamically control the cardinality of the metrics based on the cluster state without having to restart cilium agent or prometheus server. 
+
+Such issues present a unique challenge to cluster admins and developers. By creating a mechanism to manage the cardinality of Hubble metrics dynamically based on the cluster state would help in addressing the above challenges. Recent change for [Hubble Flows dynamic exporter](https://docs.cilium.io/en/latest/observability/hubble-exporter/) has inspired this proposal.
+
+## Goals
+
+* Ability to manage the cardinality of Hubble metrics dynamically.
+* Should be able to control labels and dimensions of the metrics.
+
+
+## Non-Goals
+
+* Recommendations for Prometheus configuration at ingestion time.
+
+## Proposal
+
+### Overview
+
+We propose a new config map wrapped functionality in Hubble metrics which can be used to manage the cardinality, dimensions and targets of the metrics. Drawing from [Hubble Flows dynamic exporter](https://docs.cilium.io/en/latest/observability/hubble-exporter/) which uses a standalone config map to manage exporting of flows, we propose a similar mechanism. 
+
+Existing behavior of reading cilium-config for `hubble-metrics` be preserved for backward compatibility. 
+
+A new config called `enable-dynamic-hubble-metrics` be introduced which when set to true, will enable the dynamic management of Hubble metrics. When this config is set to true, the cilium agent will read the `hubble-metrics-config` config map to get the configuration for the metrics. When this config is set to false, the cilium agent will read the `cilium-config` for `hubble-metrics` configuration.
+
+### Config Map
+
+A new config map called `hubble-metrics-config` will be introduced which will have the following fields:
+
+- `enabled`: Boolean value to enable/disable the dynamic management of Hubble metrics.
+- `metrics`: List of metrics to be managed. Each metric will have the following fields:
+  - `name`: Name of the metric.
+  - `contextOptions`: List of [context options](https://docs.cilium.io/en/latest/observability/metrics/#context-options) to be applied to the metric.
+  - `includeFilters` : List of include flow filters to be applied to this particular metric.
+  - `excludeFilters` : List of exclude flow filters to be applied to this particular metric.
+
+### Example
+    
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+    name: hubble-metrics-config
+    namespace: kube-system
+data:
+    hubblemetrics.yaml: |
+        hubbleMetrics:
+            enabled: true
+            metrics:
+            - name: flow
```

Comment:
Awesome to see this being addressed, its a major pain point right now and I was just asking about it on [slack](https://cilium.slack.com/archives/CQRL1EPAA/p1712692929740089). Happy to help if there's some way I can be useful.

One thing I would love to see would be different levels of granularity for each of the metrics (possibly appended as a suffix to the metric name) where each granularity could have it's own set of context options. This would avoid having to opt out completely from visibility into a source/destination via excludes as you could use labels with a coarser granularity (like app) to reduce the cardinality. The annotation driven suggestion below would make a lot of sense for this kind of scheme.

Reply:
@sgargan if i understand you correctly, is your scenario to have multiple granularities for a given metrics based on the pod combinations ? Or different metrics with different context options (this is supported). 


Example:

```
 - name: flow
    -   contextOptions: 
            - sourceEgressContext=pod
            - destinationIngressContext=pod
        includeFilters: 
            - [namespace1/]
  - contextOptions:
            - sourceEgressContext=workload
            - destinationIngressContext=workload
        includeFilters: 
            - [test/]
 
```

If it is the former of having same metric like (as shown above) `flow` generating different labels based on a combination of contextOption + filter then I will defer the possibility to @chancez  or whoever knows Hubble code in depth, if this is possible and is a desired scenario ?

Reply:
I'm not entirely sure if it's a desired scenario, but with the current proposal and the fact that we're introducing structure to the `contextOptions`, it should be possible to these fields/behavior later, if it makes sense. I don't think we should block on adding new functionality unless the current proposal prohibits the possibility of adding it later, so I think we're fine here.